Source/Particles/Collision/BinaryCollision/PhotonPhoton/TwoProductUtil.H

```diff
@@ -0,0 +1,142 @@
+
+#include <cmath>
+#include <limits>
+
+namespace {
+    /**
+     * \brief Given the momenta of two colliding macrophotons in a two-photon collision,
+     * this function computes the momenta of the two product macroparticles (electron and positron).
+     *
+     * This is done by using the conservation of energy and momentum,
+     * and by assuming isotropic emission of the products in the center-of-momentum frame
+     *
+     * @param[in] u1x_in normalized momentum of the first colliding macrophotons along x (in m.s^-1)
+     * @param[in] u1y_in normalized momentum of the first colliding macrophotons along y (in m.s^-1)
+     * @param[in] u1z_in normalized momentum of the first colliding macrophotons along z (in m.s^-1)
+     * @param[in] u2x_in normalized momentum of the second colliding macrophotons along x (in m.s^-1)
+     * @param[in] u2y_in normalized momentum of the second colliding macrophotons along y (in m.s^-1)
+     * @param[in] u2z_in normalized momentum of the second colliding macrophotons along z (in m.s^-1)
+     * @param[out] u1x_out normalized momentum of the first product macroparticles along x (in m.s^-1)
+     * @param[out] u1y_out normalized momentum of the first product macroparticles along y (in m.s^-1)
+     * @param[out] u1z_out normalized momentum of the first product macroparticles along z (in m.s^-1)
+     * @param[out] u2x_out normalized momentum of the second product macroparticles along x (in m.s^-1)
+     * @param[out] u2y_out normalized momentum of the second product macroparticles along y (in m.s^-1)
+     * @param[out] u2z_out normalized momentum of the second product macroparticles along z (in m.s^-1)
+     * @param[in] engine the random engine (used to calculate the angle of emission of the products)
+     */
+    AMREX_GPU_HOST_DEVICE AMREX_INLINE
+    void TwoProductComputeProductMomenta (
+                            const amrex::ParticleReal& u1x_in,
+                            const amrex::ParticleReal& u1y_in,
+                            const amrex::ParticleReal& u1z_in,
+                            const amrex::ParticleReal& u2x_in,
+                            const amrex::ParticleReal& u2y_in,
+                            const amrex::ParticleReal& u2z_in,
+                            amrex::ParticleReal& u1x_out,
+                            amrex::ParticleReal& u1y_out,
+                            amrex::ParticleReal& u1z_out,
+                            amrex::ParticleReal& u2x_out,
+                            amrex::ParticleReal& u2y_out,
+                            amrex::ParticleReal& u2z_out,
+                            const amrex::RandomEngine& engine )
+    {
+        using namespace amrex::literals;
+
+        constexpr amrex::ParticleReal c_sq = PhysConst::c * PhysConst::c;
+        constexpr amrex::ParticleReal inv_csq = 1._prt / ( c_sq );
+        constexpr amrex::ParticleReal me = PhysConst::m_e;
+        auto constexpr pow2 = [](double const x) { return x*x; };
```

Comment:
I would suggest changing the input parameter to auto, unless there is a strong reason to use a `double` here
```suggestion
        auto constexpr pow2 = [](auto const x) { return x*x; };
```

Reply:
I think that it could be useful to have these `pow2`, `pow4` functions implemented in a utility file somewhere (maybe in `ablastr`?). Using some template metaprogramming it could be even possible to define a `pow<N>(x)` function.
Anyway, I would not suggest doing that in this PR!

Reply:
@aeriforme , you can now use: `amrex::Math::powi<N>(x)`
https://github.com/AMReX-Codes/amrex/blob/development/Src/Base/AMReX_Math.H